katcp/inspecting_client.py

```diff
@@ -21,6 +22,70 @@
 ic_logger = logging.getLogger("katcp.inspect_client")
 RequestType = namedtuple('Request', 'name description')
 
+class ExponentialRandomBackoff(object):
+
+    def __init__(
+            self, delay_initial=1., delay_max=90., exp_fac=3., randomicity=0.95):
+        """Calculate random retry timeouts that increase exponentially
+
+        Input Parameters
+        ----------------
+
+        delay_initial : float
+            Initial base delay in seconds
+        delay_max : float
+            Maximum delay in seconds
+        exp_fac : float
+            Increase the base timeout by this factor for each failure
+        randomicity : float
+            Fraction of timeout that should be randomly calculated. If
+            randomicity is 0, the base delay will always be used, if it is 1, a
+            random value between 0 and the base delay is calculated, or the
+            weighted average of the two for intermediate values.
+
+        """
+        self.delay_initial = delay_initial
+        self.delay_max = delay_max
+        self.exp_fac = exp_fac
+        """Increase timeout by this factor for each consecutive failure"""
+        self._base_delay = self.delay_initial
+        self.randomicity = randomicity
+        self._update_delay()
+
+    def _update_delay(self):
+        r = self.randomicity
+        assert 0 <= r <= 1
+        assert 0 <= self.delay_initial <= self.delay_max
```

Comment:
I've refactored the validation, source code comment should hopefully explain the idea.